text/0000-multi-part-examples-in-rustdoc/index.md

```diff
@@ -0,0 +1,231 @@
+
+- Already, a rust code example is typically a nested embedding:
+  a rust file; in which there's a markdown comment; in which there's a rust code block.
+  This proposal would add another layer of nesting to this.
+  Another layer of nesting would incur mental overhead and increased implementation cost for tooling compared with the suggested approach.
+- For longer nested comments, might be cumbersome.
+  Imagine comments that include code blocks. That's another nesting. Imagine long text. You'd have to handle line breaks and comment slashes (`//`).
+
+Advantages:
+
+- Seems rather futuristic and perhaps could be part of a larger langauge change, where all comments are by definition markdown, and therefore GUI editors could display rendered versions of comments in place of the actual comments.
+- There's no splitting of code blocks.
+
+We imagine a more convenient authoring experience with the suggested design.
+Even though it feels like a close call, that tips the scale.
+
+Also, the suggested approach is not mutually exclusive with this approach, so that this approach could remain a future possibility.
+
+## Tree variant
+
+This design is similar to the suggested design, but blocks may form tree structures. Multiple blocks may continue from a same prior block.
+
+The advantage is that it would allow code sharing between multiple examples.
+
+The disadvantage is that the result might not be reader-friendly.
+Documentation is pages that are consumed usually top to bottom.
+Adding code blocks that continue one another in some tree structure may seem like a good idea from the perspective of the developer/author but would probably not actually be nice for the reader.
+
+## Interleaved variant
+
+This is a feature that is additional to the suggested design, where the blocks of a multi-part example may interleave with blocks of other multi-part examples.
+
+We feel that this could lead to confusing documentation.
+
+## Q: What is the impact of not doing this?
+
+If this feature is not implemented and no other solution is implemented for including text in long code examples, then long code examples in rust documentation continue to be inconvenient for both authors and readers.
+
+# Prior art
+[prior-art]: #prior-art
+
+This RFC is the evolution of a comment analyzing the existing [RFC 2894][rfc_2894]. It follows [this comment][rfc_comment].
+
+If we look at how programming books are commonly formatted, we will often find multi-part code examples.
+They are split into multiple code blocks and continue each other linearly.
+Between the code blocks is usually verbose educational text.
+Rust documentation authors already seem to follow this style while using one or more of the workarounds described earlier.
+This RFC officialy adopts this style.
+
+One could claim that this RFC allows code examples to be written in [literate programming][literate_programming] style.
+
+We are not familiar with a feature equivalent to the doc-tests in other languages.
```

Comment:
Doctests are pretty common.

- [Python](https://docs.python.org/3/library/doctest.html)
- [Ruby](https://www.rubydoc.info/gems/doctest-rspec/0.0.3)
- [Elixir](https://elixir-lang.org/getting-started/mix-otp/docs-tests-and-with.html#doctests)
- [D](https://dlang.org/spec/unittest.html#documented-unittests)
- Racket has runtime contracts that are part of the documentation: https://blog.racket-lang.org/2012/06/submodules.html#in-source-documentation.

It would be good to know if any of them have interleaved tests like this.

Reply:
Thank you for the links to doc-tests in other languages. Actually, our paragraph wasn't clear.
We know that doc-tests is a common feature.
What we were trying to say is that we're not aware of are features in other languages that are equivalent to the one we are suggesting in this RFC.
Corrected with 9992d9f2eae4c4c38c170d2d20d575c6ca53dd8f.

We hope that the community could refer us to any of those.
Note that for an equivalent feature to be really equivalent, it needs to:

1. Generate readable documentation, such as HTML from source code documentation
2. Extract tests from examples in that source code documentation
3. Support some rich text format interleaved with code lines of those examples

Reply:
> It would be good to know if any of them have interleaved tests like this.

This is supported in Python where you can write a docstring like this:

```python
"""
First code block:

>>> x = 10

Second code block:

>>> x
10
"""
```